### PR TITLE
Add orphacodes to diseases

### PR DESCRIPTION
**Pull request by TereseBo**
This PR adds diseases annotated with hpo-terms and hgnc-ids to the disease_terms in the scout database.
<details>

<summary>Testing on cg-vm1 server (Clinical Genomics Stockholm)</summary>

**Prepare for testing**
1. Make sure the PR is pushed and available on [Docker Hub](https://hub.docker.com/repository/docker/clinicalgenomics/scout-server-stage)
1. Fist book your testing time using the Pax software available at [https://pax.scilifelab.se/](https://pax.scilifelab.se). The resource you are going to call dibs on is `scout-stage` and the server is `cg-vm1`.
1. `ssh <USER.NAME>@cg-vm1.scilifelab.se`
1. `sudo -iu hiseq.clinical`
1. `ssh localhost`
1. (optional) Find out which scout branch is currently deployed on cg-vm1: `podman ps`
1. Stop the service with current deployed branch: `systemctl --user stop scout.target`
1. Start the scout service with the branch to test: `systemctl --user start scout@<this_branch>`
1. Make sure the branch is deployed: `systemctl --user status scout.target`
1. After testing is done, repeat procedure at [https://pax.scilifelab.se/](https://pax.scilifelab.se), which will release the allocated resource (`scout-stage`) to be used for testing by other users.
</details>

<details>
<summary>Testing on hasta server (Clinical Genomics Stockholm)</summary>

**Prepare for testing**
1. `ssh <USER.NAME>@hasta.scilifelab.se`
1. Book your testing time using the Pax software. `us; paxa -u <user> -s hasta -r scout-stage`. You can also use the WSGI Pax app available at [https://pax.scilifelab.se/](https://pax.scilifelab.se).
1. (optional) Find out which scout branch is currently deployed on cg-vm1: `conda activate S_scout; pip freeze | grep scout-browser`
1. Deploy the branch to test: `bash /home/proj/production/servers/resources/hasta.scilifelab.se/update-tool-stage.sh -e S_scout -t scout -b <this_branch>`
1. Make sure the branch is deployed: `us; scout --version`
1. After testing is done, repeat the `paxa` procedure, which will release the allocated resource (`scout-stage`) to be used for testing by other users.
</details>

**How to test**:
Setup and update diseases should now result in ORPHA:codes being included in the database.

Make sure both ORPHA and OMIM terms with HPO-terms and genes ( if applicable) are included in the database disease_terms. The fields in the database should be the same as before the change. 

**Expected outcome**:
Both the number of disease_terms and the information included in some of these (HPO-terms and genes) should be increased.

**Review:**
- [ ] code approved by
- [ ] tests executed by

### CODE REVIEW

**Comment by northwestwitch** [comment]
What happened with this PR, why did you close it?

**Comment by TereseBo** [comment]
Because I realised I had broken A LOT of tests and was not, in fact almost done >,<

**Review by northwestwitch — COMMENTED** [review]
Let's do like last time, testing first and code review later. 

I've done some tests on my local computer and the first thing I noticed is that loading diseases with this branch introduces a **change in the `disease_nr` field of the documents**. Before it was an integer and now it becomes a string.

Did you change it like this on purpose and why? I don't think this is a good idea, because the old format is what is stored in the database when you associate a diagnosis to a case, see for instance demo case:

image


This change might affect the functionality of pages in unpredicted ways I think.

Another thing I've noticed is that if you run `scout setup demo` with this branch you get 5000+  ORPHA codes loaded, are you sure you are loading just terms from the demo resources?

Another change that might potentially disrupt the update diseases functionality in other facilities: if you run scout update diseases -f <path_to_downloaded_resources> and the ORPHA files are not there then the update terms fails. This is something we handle differently at the moment I think. If the files are not there it should try to download them and then fail if it can't. I think this is what happens with OMIM at least?

**Comment by TereseBo** [comment]
I will change the disease_nr to int, is the case only  linked to the disease by the number?  Is it likely this might cause future issues? Currently any issues due to this is prevented by the changes made to the adapter functions to only retrieve OMIM diseases?

I have added the demo files from Orphadata to the demo setup and also included them in the setup command. 

I have uppdated the update diseases command so it runs when files are missing from the folder. I tried to do the same implementation as for the hpo_annotations from the phenotype.hpoa file. It the information is not supplied to the load_diseases_function it will be downloaded during it's execution. Perhaps a try/except and an information log should be added there tor prevent crashing if it fails?

**Comment by northwestwitch** [comment]
> is the case only linked to the disease by the number? Is it likely this might cause future issues? 

I don't know. We should inspect the code and change if this is the case using the OMM:number or ORPHA:number.

> Currently any issues due to this is prevented by the changes made to the adapter functions to only retrieve OMIM diseases?

I think so, but I am not 100% sure, so we should test well before merging this PR

> I have added the demo files from Orphadata to the demo setup and also included them in the setup command.

Nice! 💯 

 > Perhaps a try/except and an information log should be added there tor prevent crashing if it fails?

For the HPO or the diseases? An abort exit with a meaningful message would be nice

**Comment by northwestwitch** [comment]
I have tested again with main now, and actually if you don't provide a folder with files then it downloads the OMIM files (given that you pass the OMIM api key). If you point to a folder with no files it aborts. So it's fine to keep it like this if you want!

File: tests/parse/test_parse_disease_terms.py
Comment (northwestwitch) [inline tests/parse/test_parse_disease_terms.py:40]: This is the perfect situation where you can parametrize the test into 2 distinct tests using @pytest.mark.parametrize, alternatively I would split this tests into 2 single tests, one for OMIM and one for orpha

File: tests/parse/test_parse_disease_terms.py
Comment (northwestwitch) [inline tests/parse/test_parse_disease_terms.py:113]:
```suggestion
    for term in disease["hpo_terms"]:
       assert "HP:" in term
```

That check was very specific 😀

File: tests/parse/test_parse_hpo_mappings.py
Comment (northwestwitch) [inline tests/parse/test_parse_hpo_mappings.py:17]: 💯

File: tests/parse/test_parse_orpha.py
Comment (northwestwitch) [inline tests/parse/test_parse_orpha.py:10]:
```suggestion
def test_parse_orpha_downloads(orphadata_en_product6_lines: List[str]):
```

File: tests/parse/test_parse_orpha.py
Comment (northwestwitch) [inline tests/parse/test_parse_orpha.py:11]: Provide a docstring for this function

File: tests/parse/test_parse_orpha.py
Comment (northwestwitch) [inline tests/parse/test_parse_orpha.py:20]:
```suggestion
def test_get_orpha_to_genes_information(orphadata_en_product6_lines: List[str]):
```

File: tests/parse/test_parse_orpha.py
Comment (northwestwitch) [inline tests/parse/test_parse_orpha.py:24]:
```suggestion
    disease: dict = result["ORPHA:585"]
```

File: tests/parse/test_parse_orpha.py
Comment (northwestwitch) [inline tests/parse/test_parse_orpha.py:21]: doctring

File: tests/parse/test_parse_orpha.py
Comment (northwestwitch) [inline tests/parse/test_parse_orpha.py:32]:
```suggestion
def test_get_orpha_to_hpo_information(orphadata_en_product4_lines: List[str]):
```

File: tests/parse/test_parse_orpha.py
Comment (northwestwitch) [inline tests/parse/test_parse_orpha.py:33]: docstring

**Review by northwestwitch — CHANGES_REQUESTED** [review]
Fantastic job! Code is going to be much more clear now, and I really like the new functions that parse/fix both OMIM and ORPHA terms. I have a lot of comments, but don't worry, it's mostly very easy fixes concerning type hints and whatnot.

I haven't had time yet to run the command and check the result in the database, but I'll do it ASAP.

File: scout/build/disease.py
Comment (TereseBo) [reply scout/build/disease.py:30]: I have readded the comment but I must admit I'm not completely clear on what iOS happening or why. If the gene symbol is found in alias genes and has a "true" id that is added to the list of hence_ids and if it does not have a "true" id, all id:s are added?

File: tests/parse/test_parse_disease_terms.py
Comment (TereseBo) [reply tests/parse/test_parse_disease_terms.py:113]: Isn't what we actually want to check if the diseases are given the correct terms? Perhaps I'm overly cautious >,<.

File: tests/parse/test_parse_disease_terms.py
Comment (TereseBo) [reply tests/parse/test_parse_disease_terms.py:42]: The test is for a function combining orpha and omim terms and asserts the contents from both are present in the result

File: scout/commands/setup/setup_scout.py
Comment (northwestwitch) [inline scout/commands/setup/setup_scout.py:101]:
```suggestion
        "Path to file with map from HPO disease (OMIM and ORPHA) to phenotype term with annotation. This is the file  called phenotype.hpoa"
    ),
```

File: scout/constants/phenotype.py
Comment (northwestwitch) [reply scout/constants/phenotype.py:37]: I would still rename these files so it's clearer what these lines are

File: scout/constants/phenotype.py
Comment (northwestwitch) [reply scout/constants/phenotype.py:41]: Same here

File: scout/demo/resources/__init__.py
Comment (northwestwitch) [reply scout/demo/resources/__init__.py:56]: Same

File: scout/demo/resources/__init__.py
Comment (northwestwitch) [reply scout/demo/resources/__init__.py:57]: Same

File: scout/load/disease.py
Comment (northwestwitch) [inline scout/load/disease.py:17]:
```suggestion
    genemap_lines: List[str],
```

File: tests/build/test_build_disease.py
Comment (northwestwitch) [inline tests/build/test_build_disease.py:24]:
```suggestion
def test_build_disease_wrong_value(test_disease, test_disease_id):
```

File: tests/build/test_build_disease.py
Comment (northwestwitch) [inline tests/build/test_build_disease.py:29]:
```suggestion
```

**Review by northwestwitch — COMMENTED** [review]
Good stuff! 

I think you missed some of my comments in scout/parse/disease_terms.py but by the point of view of the code I think this PR is fine. What I think we should do before merging it:

- Make sure that all the queries in the code where OMIM terms are collected are fixed so only terms with source=OMIM are returned
- Test it on hasta stage, to see how many terms (both OMIM and ORPHA) are loaded into the database, and the difference with the OMIM terms loaded yesteday in the prod database.

I'll do the terms uploading on hasta stage now

**Comment by northwestwitch** [comment]
I've been trying to update diseases on hasta stage with the files I've downloaded and used yesterday for updating the database and got the following error:

```
[chiara.rasi@hasta:~] [S_base] 1m50s $ scout --config /home/proj/stage/servers/config/hasta.scilifelab.se/scout-stage.yaml update diseases -f scout_downloads_20-12-2023/
2023-12-21 08:54:18 hasta.scilifelab.se scout.commands.base[196783] INFO Running scout version 4.75
2023-12-21 08:54:18 hasta.scilifelab.se scout.commands.base[196783] DEBUG Debug logging enabled.
2023-12-21 08:54:18 hasta.scilifelab.se scout.commands.base[196783] WARNING Support for launching Scout using a .yaml config file is deprecated and will be removed in the next major release (5.0). Use a PYTHON (.py) config file instead.
2023-12-21 08:54:18 hasta.scilifelab.se scout.adapter.client[196783] INFO Connected to MongoDB 7.0.2
2023-12-21 08:54:18 hasta.scilifelab.se scout.adapter.mongo.base[196783] DEBUG Database name: scout-stage
2023-12-21 08:54:18 hasta.scilifelab.se scout.server.app[196783] INFO Enable Phenopacket API
2023-12-21 08:54:20 hasta.scilifelab.se scout.adapter.mongo.hgnc[196783] INFO Fetching all genes by alias
2023-12-21 08:54:20 hasta.scilifelab.se scout.parse.orpha[196783] INFO Parsing Orphadata en_product6
Traceback (most recent call last):
  File "/home/proj/stage/bin/miniconda3/envs/S_scout/bin/scout", line 8, in <module>
    sys.exit(cli())
  File "/home/proj/stage/bin/miniconda3/envs/S_scout/lib/python3.7/site-packages/click/core.py", line 1130, in __call__
    return self.main(*args, **kwargs)
  File "/home/proj/stage/bin/miniconda3/envs/S_scout/lib/python3.7/site-packages/click/core.py", line 1055, in main
    rv = self.invoke(ctx)
  File "/home/proj/stage/bin/miniconda3/envs/S_scout/lib/python3.7/site-packages/click/core.py", line 1657, in invoke
    return _process_result(sub_ctx.command.invoke(sub_ctx))
  File "/home/proj/stage/bin/miniconda3/envs/S_scout/lib/python3.7/site-packages/click/core.py", line 1657, in invoke
    return _process_result(sub_ctx.command.invoke(sub_ctx))
  File "/home/proj/stage/bin/miniconda3/envs/S_scout/lib/python3.7/site-packages/click/core.py", line 1404, in invoke
    return ctx.invoke(self.callback, **ctx.params)
  File "/home/proj/stage/bin/miniconda3/envs/S_scout/lib/python3.7/site-packages/click/core.py", line 760, in invoke
    return __callback(*args, **kwargs)
  File "/home/proj/stage/bin/miniconda3/envs/S_scout/lib/python3.7/site-packages/click/decorators.py", line 26, in new_func
    return f(get_current_context(), *args, **kwargs)
  File "/home/proj/stage/bin/miniconda3/envs/S_scout/lib/python3.7/site-packages/flask/cli.py", line 357, in decorator
    return __ctx.invoke(f, *args, **kwargs)
  File "/home/proj/stage/bin/miniconda3/envs/S_scout/lib/python3.7/site-packages/click/core.py", line 760, in invoke
    return __callback(*args, **kwargs)
  File "/home/proj/stage/bin/miniconda3/envs/S_scout/lib/python3.7/site-packages/scout/commands/update/disease.py", line 99, in diseases
    orpha_to_hpo_lines=resources["orphadata_en_product4_lines"],
  File "/home/proj/stage/bin/miniconda3/envs/S_scout/lib/python3.7/site-packages/scout/load/disease.py", line 39, in load_disease_terms
    orpha_to_genes_lines=orpha_to_genes_lines,
  File "/home/proj/stage/bin/miniconda3/envs/S_scout/lib/python3.7/site-packages/scout/parse/disease_terms.py", line 21, in get_all_disease_terms
    orpha_to_hpo_lines=orpha_to_hpo_lines, orpha_to_genes_lines=orpha_to_genes_lines
  File "/home/proj/stage/bin/miniconda3/envs/S_scout/lib/python3.7/site-packages/scout/parse/disease_terms.py", line 130, in get_orpha_disease_terms
    orpha_disease: Dict = get_orpha_to_genes_information(lines=orpha_to_genes_lines)
  File "/home/proj/stage/bin/miniconda3/envs/S_scout/lib/python3.7/site-packages/scout/parse/orpha.py", line 41, in get_orpha_to_genes_information
    for gene_association in gene_list:
TypeError: 'NoneType' object is not iterable
θ62° [chiara.rasi@hasta:~] [S_base] 29s 1 $
```

**Comment by northwestwitch** [comment]
Comparison between prod and stage DBs after loading diseases using this branch:

Prod has 9008 OMIM terms:

image

<hr>

Stage has the same number of OMIM terms: 🆗 

image

<hr>

Plus 5800k + ORPHA terms: 🆗 

image

**Review by northwestwitch — APPROVED** [review]
I am approving now since the code looks OK and works on stage. I think it's OK to merge when you feel ready with the fixes of this PR and the stuff in #4306 is addressed.

Good job! 🥇

**Comment by TereseBo** [comment]
Resources were downloaded with scout download everything --api-key <api-key> --out-dir tb240115 and then database was updated with  scout update diseases -f ./tb240115 resulting in the scout-stage database to be loaded with 14786 disease terms.

image

Screenshot 2024-01-15 at 12 27 58


The same commands on main branch loaded 9015 disease terms

Screenshot 2024-01-15 at 12 56 24

**Comment by TereseBo** [comment]
Tested also on cg-vm1. Disease_terms loaded in database but not visible on 

In scout-stage database ORPHA terms are present: 
Screenshot 2024-01-15 at 14 55 18

But these diseases are not visible in https://scout-stage.scilifelab.se/diagnoses:
Screenshot 2024-01-15 at 14 49 57
 
And can not be found when searching to add a disease to a case (OMIM terms can still be found)::
Screenshot 2024-01-15 at 14 55 05

And the ORPHA terms in the database are not listed when going to the page of a linked gene (https://scout-stage.scilifelab.se/genes/5173):
Screenshot 2024-01-15 at 15 01 11

**Pull request merged**
